### PR TITLE
Path optimization

### PR DESCRIPTION
**Pull request by AnnaDaiZH**
- add intermediate points to generated OMPL path
- add IG yaw optimisation to path vertices

### CODE REVIEW

File: se_ompl/include/se/utils/ompl_to_eigen.hpp [hunk 1]
```diff
@@ -105,9 +105,9 @@ class OmplToEigen {
   // state [meter] to voxel
   static Eigen::Vector3i convertState_v(const ompl::base::State &state, const float dim) {
 
-    Eigen::Vector3i eigen_point(static_cast<int>(state.as<ompl::base::RealVectorStateSpace::StateType>()->values[0]/dim),
-                                static_cast<int>(state.as<ompl::base::RealVectorStateSpace::StateType>()->values[1]/dim),
-                                static_cast<int>(state.as<ompl::base::RealVectorStateSpace::StateType>()->values[2]/dim));
+    Eigen::Vector3i eigen_point(std::round(state.as<ompl::base::RealVectorStateSpace::StateType>()->values[0]/dim),
+                                std::round(state.as<ompl::base::RealVectorStateSpace::StateType>()->values[1]/dim),
+                                std::round(state.as<ompl::base::RealVectorStateSpace::StateType>()->values[2]/dim));
```
Comment (sotpapathe) [inline se_ompl/include/se/utils/ompl_to_eigen.hpp]: The cast to int is the way supereight currently converts from map to voxel coordinates. Why is round needed in this case?

File: se_shared/include/se/path_planning/exploration_utils.hpp [hunk 2]
```diff
@@ -223,10 +223,10 @@ static inline Eigen::Vector3i toVoxelCoord(const Eigen::Vector3f &coord_m, const
 static inline bool boundHeight(int *h, const float h_max, const float h_min, const float res) {
 
   if (*h > h_max / res) {
-    *h = static_cast<int>(h_max / res) - 0.2;
+    *h = static_cast<int>((h_max - 0.25f) / res);
     return true;
   } else if (*h < h_min / res) {
-    *h = static_cast<int>(h_min / res) + 0.2;
+    *h = static_cast<int>((h_min +0.25f )/ res) ;
```
Comment (sotpapathe) [inline se_shared/include/se/path_planning/exploration_utils.hpp]: Why clamp the height +-0.25m above/below the minimum/maximum? Is this the safety radius? If so, it should probably be parameterized.

File: se_ompl/include/se/path_planner_ompl.hpp [hunk 3]
```diff
@@ -437,15 +446,16 @@ template<typename FieldType>
...
   lower_bound_ = lower_bound_v_.cast<float>() * dim;
   upper_bound_ = upper_bound_v_.cast<float>() * dim;
   bounds.setLow(0, lower_bound_v_.x() * dim - buffer_m);
   bounds.setLow(1, lower_bound_v_.y() * dim - buffer_m);
-  bounds.setLow(2, planning_params_.height_min - 0.2);
+  bounds.setLow(2,  ground_height_);
   bounds.setHigh(0, upper_bound_v_.x() * dim + buffer_m);
   bounds.setHigh(1, upper_bound_v_.y() * dim + buffer_m);
-  bounds.setHigh(2, planning_params_.height_max + 0.2);
+  bounds.setHigh(2, ground_height_ +2.8f);
```
Comment (sotpapathe) [inline se_ompl/include/se/path_planner_ompl.hpp]: What does ground_height_ represent? Why the 2.8m offset? Wouldn't it be simpler to have an environment height and set the z limits to 0 and environment_height_ (e.g. 3 m)?

File: se_denseslam/include/se/boundary_extraction.hpp [hunk 4]
```diff
@@ -117,27 +121,41 @@ static inline void getFreeMapBounds(const std::shared_ptr<se::Octree<T> > octree
...
+
+    // }
+    // if (upper_bound_tmp.norm() > upper_bound.norm() && valid_upper) {
+    //   // std::cout << "upper_bound from " << upper_bound.format(InLine) << " to "
+    //   // << upper_bound_tmp.format(InLine) << std::endl;
+    //   upper_bound = upper_bound_tmp;
+    // }
+
   }
   upper_bound += Eigen::Vector3i(8, 8, 8);
```
Comment (sotpapathe) [inline se_denseslam/include/se/boundary_extraction.hpp]: Remove this line if not done already.

File: se_denseslam/include/se/boundary_extraction.hpp [hunk 5]
```diff
@@ -117,27 +121,41 @@ static inline void getFreeMapBounds(const std::shared_ptr<se::Octree<T> > octree
...
+      lower_bound.z() = lower_bound_tmp.z() < lower_bound.z() ? lower_bound_tmp.z() : lower_bound.z();
     }
-    if (upper_bound_tmp.norm() > upper_bound.norm() && valid_upper) {
-      // std::cout << "upper_bound from " << upper_bound.format(InLine) << " to "
-      // << upper_bound_tmp.format(InLine) << std::endl;
-      upper_bound = upper_bound_tmp;
+    if (valid_upper){
+      upper_bound.x() = upper_bound_tmp.x() > upper_bound.x() ? upper_bound_tmp.x() : upper_bound.x();
+      upper_bound.y() = upper_bound_tmp.y() > upper_bound.y() ? upper_bound_tmp.y() : upper_bound.y();
+      upper_bound.z() = upper_bound_tmp.z() > upper_bound.z() ? upper_bound_tmp.z() : upper_bound.z();
```
Comment (sotpapathe) [inline se_denseslam/include/se/boundary_extraction.hpp]:
```
// Add 1 to the the voxel side length into account.
if (upper_bound_tmp.x() + 1 > upper_bound.x())
    upper_bound.x() = upper_bound_tmp.x() + 1;
```
same for y and z.

File: se_ompl/include/se/path_planner_ompl.hpp (hunk 3 above)
Comment (AnnaDaiZH) [reply se_ompl/include/se/path_planner_ompl.hpp]: you don't want the drone to crash into the ceiling

Comment (AnnaDaiZH) [reply se_ompl/include/se/path_planner_ompl.hpp]: adding new param ceiling_height

**Review by sotpapathe — APPROVED** [review]

**Pull request merged**